### PR TITLE
feat(diagnostic): add "code" to the diagnostic structure

### LINKED ISSUE #17260: LSP: Standard diagnostic code in the diagnostic structure
### Feature already in Vim?

No, it's about neovim API

### Feature description

I propose neovim's diagnostic structure (introduced in neovim 0.6.0+) should include `code` as a standard entry. This can be error/warning code from many LSP or linter implementations (such as eslint, pylint, flake8, etc.).

Currently, error codes are stored under either:
- (1) `diagnostic.user_data.lsp.code` (neovim's current way of handling this, see `:help diagnostic-table`) or 
- (2) `diagnostic.code` as a top-level entry (e.g., [null-ls.nvim#632](https://github.com/jose-elias-alvarez/null-ls.nvim/issues/632)) which is "non-standard".

My proposal is to make `code` standarized and become a top-level entry in the diagnostic table.

### More information and context

Continuation from discussion https://github.com/jose-elias-alvarez/null-ls.nvim/issues/632#issuecomment-1026093907.

[The Diagnostic API defined in LSP](https://microsoft.github.io/language-server-protocol/specification#diagnostic) includes `code` (as well as `source`, `message`, `severity`, etc.) as a top-level entry. Neovim's diagnostic data structure does not, as per `:help diagnostic-table`.

```
A diagnostic is a Lua table with the following keys:                        
                                                                            
        bufnr: Buffer number                                                
        lnum: The starting line of the diagnostic                           
        end_lnum: The final line of the diagnostic                          
        col: The starting column of the diagnostic                          
        end_col: The final column of the diagnostic                         
        severity: The severity of the diagnostic vim.diagnostic.severity    
        message: The diagnostic text                                        
        source: The source of the diagnostic                                
        user_data: Arbitrary data plugins or users can add                  
```

When the API was introduced in neovim 0.6.0, `code` were not included as a standard field of the diagnostic table structure, but I believe this is worth including and there is no reason to exclude that. Error codes are very common in diagnostic and linters, and many existing plugins such as [trouble.nvim](https://github.com/folke/trouble.nvim) are displaying the error codes.

Typical LSP servers configured (but not all) through nvim-lspconfig would follow the `user_data.lsp.code` convention, as seen in a helper function [`diagnostic_lsp_to_vim`](https://github.com/neovim/neovim/blob/v0.6.0/runtime/lua/vim/lsp/diagnostic.lua#L92-L118). But some other diagnostic sources such as [null-ls.nvim](https://github.com/jose-elias-alvarez/null-ls.nvim), not necessarily being aware of LSP, would want to use different namespaces. @jose-elias-alvarez, the author of null-ls.nvim proposed `user_data.null_ls` in jose-elias-alvarez/null-ls.nvim#632, but this would make use of the exposed error code quite complicated for the clients.

For instance, some plugins like [trouble.nvim](https://github.com/folke/trouble.nvim/blob/aae12e7b23b3a2b8337ec5b1d6b7b4317aa3929b/lua/trouble/util.lua#L177) will fail to extract the data because of the plugin-specific namespace. In addition, picking up the code information for custom diagnostic messages (e.g., `vim.diagnostic.config { float = { message = function(diagnostic) ... end } }`) will be also difficult unless `code` is standarized.

Any reasons why `code` were not included back then, and possible issues of this (potentially) breaking change? Thanks for your consideration.

/cc @jose-elias-alvarez @folke

### PR DESCRIPTION
**Pull request by mxteries**
This should address everything mentioned here https://github.com/neovim/neovim/issues/17260#issuecomment-1030686655. I also commented a few questions I had below.

And let me know if there's other tests I should write

Closes #17260

### CODE REVIEW

File: runtime/doc/diagnostic.txt [hunk 1]
```diff
@@ -39,16 +39,18 @@ modify the diagnostics for a buffer (e.g. |vim.diagnostic.set()|) then it
...
 	bufnr: Buffer number
-	lnum: The starting line of the diagnostic
+	lnum(*): The starting line of the diagnostic
 	end_lnum: The final line of the diagnostic
-	col: The starting column of the diagnostic
+	col(*): The starting column of the diagnostic
 	end_col: The final column of the diagnostic
 	severity: The severity of the diagnostic |vim.diagnostic.severity|
-	message: The diagnostic text
+	message(*): The diagnostic text
```
Comment (mxteries) [inline runtime/doc/diagnostic.txt:51]: Something to note: `message` is supposed to be a required field, but there's no `assert` that's checking for it like there is with `lnum` and `col`: https://github.com/neovim/neovim/blob/4c24fcb91f7740410b94638b3156ba20647df579/runtime/lua/vim/diagnostic.lua#L272-L276

So I can `vim.diagnostic.set()` a diagnostic without a message, but when I try to `vim.diagnostic.open_float()` it, the call fails. If that is indeed unintended, then I can open an issue for it. It's technically a breaking change so I'm not sure if I should address it in this PR

File: runtime/lua/vim/lsp/diagnostic.lua [hunk 2]
```diff
@@ -120,7 +122,8 @@ end
 ---@private
 local function diagnostic_vim_to_lsp(diagnostics)
   return vim.tbl_map(function(diagnostic)
-    return vim.tbl_extend("error", {
+    return vim.tbl_extend("keep", {
+      -- "keep" the below fields over any duplicate fields in diagnostic.user_data.lsp
```
Comment (mxteries) [inline runtime/lua/vim/lsp/diagnostic.lua:126]: This function looks like it's only called by deprecated `vim.lsp.diagnostic` functions. Should I just not change this function if that's the case?

File: runtime/doc/diagnostic.txt (hunk 1 above)
Comment (gpanders) [reply runtime/doc/diagnostic.txt:51]: Hmm yes I think we should assert for `diagnostic.message` as well. I don't consider that a breaking change however, as the diagnostics are already assuming that `message` exists, so omitting it already causes errors (as you've noticed), we'd just be making the error more explicit.

**Review by gpanders — APPROVED** [review]
LGTM. If you want to go ahead and add the `assert` for the message field you can, otherwise I can do it myself in a followup PR.

**Review by mfussenegger — APPROVED** [review]

**Comment by mxteries** [comment]
> LGTM. If you want to go ahead and add the `assert` for the message field you can, otherwise I can do it myself in a followup PR.

Let's just go with the PR as is then, in case adding the assert involves more work than it seems.

`make test` passes locally for me, but it looks like there's a failed build `builds.sr.ht: openbsd.yml` due to a timeout. Let me know if I need to do anything about that

**Comment by gpanders** [comment]
Thanks!

**Comment by mjlbach** [comment]
Note, some users have requested this be rendered by default in floating_preview

**Comment by neovim-discourse** [comment]
This pull request has been mentioned on **Neovim Discourse**. There might be relevant details there:

https://neovim.discourse.group/t/how-to-display-diagnostic-code/2141/3

**Pull request merged**